llpc/lower/llpcSpirvLowerConstImmediateStore.cpp

```diff
@@ -51,164 +52,178 @@ namespace Llpc {
+  SmallVector<GetElementPtrInst *> geps;
+
+  // Step 1: Determine if the alloca can be converted and find the relevant store(s)
+  SmallVector<std::pair<Value *, std::optional<uint64_t>>> pointers;
+  pointers.emplace_back(allocaInst, 0);
+  do {
+    auto [pointer, index] = pointers.pop_back_val();
+    for (Use &use : pointer->uses()) {
+      auto user = cast<Instruction>(use.getUser());
       if (auto storeInst = dyn_cast<StoreInst>(user)) {
-        if (pointer == storeInst->getValueOperand() || storeInstFound || !isOuterPointer) {
-          // Pointer escapes by being stored, or we have already found a "store"
-          // instruction, or this is a partial "store" instruction.
-          return nullptr;
+        if (pointer == storeInst->getValueOperand() || !index.has_value()) {
+          // Pointer escapes by being stored, or we store to a dynamically indexed (or otherwise complex) pointer.
+          return false;
         }
-        storeInstFound = storeInst;
-      } else if (auto getElemPtrInst = dyn_cast<GetElementPtrInst>(user))
-        pointers.push_back(getElemPtrInst);
-      else if (!isa<LoadInst>(user) && !isAssumeLikeIntrinsic(user)) {
-        // Pointer escapes by being used in some way other than "load/store/getelementptr".
-        return nullptr;
-      }
-    }
 
-    if (pointers.empty())
-      break;
+        Value *storeValue = storeInst->getValueOperand();
+        if (!isa<Constant>(storeValue))
+          return false;
 
-    pointer = pointers.back();
-    pointers.pop_back();
-    isOuterPointer = false;
-  }
+        if (storeValue->getType() == allocatedTy) {
+          if (aggregateStore || !elementStores.empty() || index.value() != 0)
+            return false;
+          aggregateStore = storeInst;
+          continue;
+        }
 
-  return storeInstFound;
-}
+        if (arrayTy && storeValue->getType() == arrayTy->getArrayElementType()) {
+          if (aggregateStore)
+            return false;
+          if (index.value() >= arrayTy->getArrayNumElements())
+            return false;
 
-// =====================================================================================================================
-// Converts an "alloca" instruction with a single constant store into a read-only global variable.
-//
-// NOTE: This erases the "store" instruction (so it will not be lowered by a later lowering pass
-// any more) but not the "alloca" or replaced "getelementptr" instruction (they will be removed
-// later by DCE pass).
-//
-// @param storeInst : The single constant store into the "alloca"
-void SpirvLowerConstImmediateStore::convertAllocaToReadOnlyGlobal(StoreInst *storeInst) {
-  auto allocaInst = cast<AllocaInst>(storeInst->getPointerOperand());
-  auto globalType = allocaInst->getAllocatedType();
-  auto initVal = cast<Constant>(storeInst->getValueOperand());
+          if (elementStores.empty())
+            elementStores.resize(arrayTy->getArrayNumElements());
```

Comment:
Is this going to cause problems if you try to compile a fuzz test that delares a local array with a billion elements?